pystac/errors.py

```diff
@@ -24,7 +24,26 @@ class STACTypeError(Exception):
     a Catalog JSON was read in as an Item.
     """
 
-    pass
+    def __init__(
+        self,
+        bad_dict: Dict[str, Any],
+        expected: type,
+        extra_message: Optional[str] = "",
+    ):
+        """
+        Construct an exception with an appropriate error message from bad_dict and the
+        expected that it didn't align with.
+
+        Args:
+            bad_dict: Dictionary that did not match the expected type
+            expected: The expected type.  This should a :class:`~pystac.STACObject`, but
+                is left as `type` for the avoidance of circular imports.
```

Comment:
Could you use:

```python
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from pystac import STACObject
```

Reply:
That would work save for some long standing [mypy hates AbstractBaseClasses](https://github.com/python/mypy/issues/4717) issues, and that ItemCollection isn't a STACObject.  So, I motion to proceed with comment brevity and leaving the annotation as `type`.